tests/unit/v1/async/test_async_transaction.py

```diff
@@ -0,0 +1,1055 @@
+        self.assertIs(wrapped.to_wrap, mock.sentinel.callable_)
+
+
+class Test__commit_with_retry(aiounittest.AsyncTestCase):
+    @staticmethod
+    @pytest.mark.asyncio
+    async def _call_fut(client, write_pbs, transaction_id):
+        from google.cloud.firestore_v1.async_transaction import _commit_with_retry
+
+        return await _commit_with_retry(client, write_pbs, transaction_id)
+
+    @mock.patch("google.cloud.firestore_v1.async_transaction._sleep")
+    @pytest.mark.asyncio
+    async def test_success_first_attempt(self, _sleep):
+        from google.cloud.firestore_v1.services.firestore import (
+            client as firestore_client,
+        )
+
+        # Create a minimal fake GAPIC with a dummy result.
+        firestore_api = mock.create_autospec(
+            firestore_client.FirestoreClient, instance=True
+        )
+
+        # Attach the fake GAPIC to a real client.
+        client = _make_client("summer")
+        client._firestore_api_internal = firestore_api
+
+        # Call function and check result.
+        txn_id = b"cheeeeeez"
+        commit_response = await self._call_fut(client, mock.sentinel.write_pbs, txn_id)
+        self.assertIs(commit_response, firestore_api.commit.return_value)
+
+        # Verify mocks used.
+        _sleep.assert_not_called()
+        firestore_api.commit.assert_called_once_with(
+            request={
+                "database": client._database_string,
+                "writes": mock.sentinel.write_pbs,
+                "transaction": txn_id,
+            },
+            metadata=client._rpc_metadata,
+        )
+
+    @mock.patch(
+        "google.cloud.firestore_v1.async_transaction._sleep", side_effect=[2.0, 4.0]
+    )
+    @pytest.mark.asyncio
+    async def test_success_third_attempt(self, _sleep):
+        from google.api_core import exceptions
+        from google.cloud.firestore_v1.services.firestore import (
+            client as firestore_client,
+        )
+
+        # Create a minimal fake GAPIC with a dummy result.
+        firestore_api = mock.create_autospec(
+            firestore_client.FirestoreClient, instance=True
+        )
+        # Make sure the first two requests fail and the third succeeds.
+        firestore_api.commit.side_effect = [
+            exceptions.ServiceUnavailable("Server sleepy."),
+            exceptions.ServiceUnavailable("Server groggy."),
+            mock.sentinel.commit_response,
+        ]
+
+        # Attach the fake GAPIC to a real client.
+        client = _make_client("outside")
+        client._firestore_api_internal = firestore_api
+
+        # Call function and check result.
+        txn_id = b"the-world\x00"
+        commit_response = await self._call_fut(client, mock.sentinel.write_pbs, txn_id)
+        self.assertIs(commit_response, mock.sentinel.commit_response)
+
+        # Verify mocks used.
+        self.assertEqual(_sleep.call_count, 2)
+        _sleep.assert_any_call(1.0)
+        _sleep.assert_any_call(2.0)
```

Comment:
Is `_sleep` here saying "wait for 1 event to finish, then wait for the 2nd event to finish" or "wait for 1 event to finish, then wait for 2 more events to finish"?

Reply:
The `_sleep` mock assert check here is saying "make sure there were two calls to sleep, where one call to sleep was 1 second and the another call to sleep was 2 seconds". `_sleep` is called in `_commit_with_retry` with multiplicatively larger intervals. The test case is checking the functionality of `_commit_with_retry` when the commit fails the first two times and succeeds the third, in which case the expected sleep is 1 second and then 2 seconds.

Reply:
Ah, okay so it's ensuring there is backoff.

Great, can we add a comment stating as much?

Reply:
Sure!